lib/dialects/abstract/query-generator.js

```diff
@@ -1493,124 +1182,273 @@ const QueryGenerator = {
+      externalAs: include.as
+    };
+    let attributes;
+    let joinQuery;
 
-    if (Array.isArray(options.order)) {
-      for (const t of options.order) {
-        if (Array.isArray(t) && _.size(t) > 1) {
-          if ((typeof t[0] === 'function' && t[0].prototype instanceof Model) || (typeof t[0].model === 'function' && t[0].model.prototype instanceof Model)) {
-            if (typeof t[t.length - 2] === 'string') {
-              validateOrder(_.last(t));
+    topLevelOptions.keysEscaped = true;
+
+    if (topLevelTableInfo.name !== parentTableName.externalAs && topLevelTableInfo.as !== parentTableName.externalAs) {
+      includeAs.internalAs = `${parentTableName.internalAs}_${include.as}`;
+      includeAs.externalAs = `${parentTableName.externalAs}.${include.as}`;
+    }
+
+    // includeIgnoreAttributes is used by aggregate functions
+    if (topLevelOptions.includeIgnoreAttributes !== false) {
+      attributes = include.attributes.map(attr => {
+        let attrAs = attr;
+        let verbatim = false;
+
+        if (Array.isArray(attr) && attr.length === 2) {
+          if (attr[0]._isSequelizeMethod) {
+            if (attr[0] instanceof Utils.Literal ||
+              attr[0] instanceof Utils.Cast ||
+              attr[0] instanceof Utils.Fn
+            ) {
+              verbatim = true;
             }
-          } else {
-            validateOrder(_.last(t));
           }
-        }
 
-        if (subQuery && (Array.isArray(t) && !(typeof t[0] === 'function' && t[0].prototype instanceof Model) && !(t[0] && typeof t[0].model === 'function' && t[0].model.prototype instanceof Model))) {
-          subQueryOrder.push(this.quote(t, model));
+          attr = attr.map($attr => $attr._isSequelizeMethod ? this.handleSequelizeMethod($attr) : $attr);
+
+          attrAs = attr[1];
+          attr = attr[0];
+        } else if (attr instanceof Utils.Literal) {
+          return attr.val; // We trust the user to rename the field correctly
+        } else if (attr instanceof Utils.Cast || attr instanceof Utils.Fn) {
+          throw new Error(
+            'Tried to select attributes using Sequelize.cast or Sequelize.fn without specifying an alias for the result, during eager loading. ' +
+            'This means the attribute will not be added to the returned instance'
+          );
         }
 
-        mainQueryOrder.push(this.quote(t, model));
+        let prefix;
+        if (verbatim === true) {
+          prefix = attr;
+        } else {
+          prefix = `${this.quoteIdentifier(includeAs.internalAs)}.${this.quoteIdentifier(attr)}`;
+        }
+        return `${prefix} AS ${this.quoteIdentifier(`${includeAs.externalAs}.${attrAs}`, true)}`;
+      });
+      if (include.subQuery && topLevelSubQuery) {
+        for (let attr of attributes) {
+          topLevelTableAttributes.subQuery.push(attr);
+        }
+      } else {
+        for (let attr of attributes) {
+          topLevelTableAttributes.main.push(attr);
```

Comment:
consider returning the attributes to the parent, make the code more functional/pure and easier to follow.